File: backend/app/api/routes/documents.py
"""项目文档树 (documents) — human/agent REST 前门(万物皆块之上的真实实现)。

文档是可编辑状态:节点内容就地更新(与追加式的聊天不同)。编辑文档就是 actor 向 AI
下达指令(save_body 会追加一个 EVENT 块)。所有路由把从 Bearer JWT 注入的 actor
(`require_auth_user`,绝不从请求体读取)传给 DocumentService,由 service 校验项目成员
身份。响应统一为 {code, message, data}。
"""

from typing import Annotated

from fastapi import APIRouter, Depends, Path, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.checker import require_auth_user
from app.auth.core import AuthUserInfo
from app.db.session import get_db
from app.domain.document.schemas import (
    CreateDocumentRequest,
    SaveBodyRequest,
    UpdateDocumentRequest,
)
from app.domain.document.services import DocumentService

router = APIRouter(tags=["Workspace · 文档"])


def _ok(data: object, *, code: int = 200, message: str = "success") -> dict:
    return {"code": code, "message": message, "data": data}


@router.get("/projects/{projectId}/documents", summary="List document tree")
async def list_documents(
    project_id: Annotated[int, Path(ge=1, alias="projectId")],
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    tree = await DocumentService(db).list_tree(project_id, auth.user_id)
    return _ok({"documents": [n.model_dump(by_alias=True) for n in tree]})


@router.post(
    "/projects/{projectId}/documents",
    summary="Create document",
    status_code=status.HTTP_201_CREATED,
)
async def create_document(
    project_id: Annotated[int, Path(ge=1, alias="projectId")],
    payload: CreateDocumentRequest,
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    doc = await DocumentService(db).create_document(
        project_id, auth.user_id, payload.title, payload.parent_id
    )
    return _ok({"document": doc.model_dump(by_alias=True)}, code=201, message="Created")


@router.get("/documents/{documentId}", summary="Get document meta + body + node tree")
async def get_document(
    document_id: Annotated[int, Path(ge=1, alias="documentId")],
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    detail = await DocumentService(db).get_document(document_id, auth.user_id)
    return _ok(detail.model_dump(by_alias=True))


@router.put("/documents/{documentId}", summary="Save document body (reconcile + event)")
async def save_document_body(
    document_id: Annotated[int, Path(ge=1, alias="documentId")],
    payload: SaveBodyRequest,
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    detail = await DocumentService(db).save_body(document_id, auth.user_id, payload.content)
    return _ok(detail.model_dump(by_alias=True))


@router.patch("/documents/{documentId}", summary="Rename / move / archive document")
async def update_document(
    document_id: Annotated[int, Path(ge=1, alias="documentId")],
    payload: UpdateDocumentRequest,
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    fields = payload.model_dump(exclude_unset=True)
    doc = await DocumentService(db).update_document(document_id, auth.user_id, fields)
    return _ok({"document": doc.model_dump(by_alias=True)})


@router.delete(
    "/documents/{documentId}",
    summary="Delete document (reparent children)",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_document(
    document_id: Annotated[int, Path(ge=1, alias="documentId")],
    auth: AuthUserInfo = Depends(require_auth_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await DocumentService(db).delete_document(document_id, auth.user_id)

File: frontend/src/network/api/documents.ts
// 知是 2.0 项目文档 API — 人类前门（REST，`/api` base，`{code,message,data}` 信封）。
// 文档即节点（Notion/飞书式）：任意文档可有子文档，树由后端直接嵌套返回。
// 复用应用统一的 axios 实例（NewApiInstance），与其它人类前门模块一致；
// 不使用 connectorFetch（那是 connector 面板专用）。

import { NewApiInstance } from './index'

// 文档类型（后端 docType 枚举，目前只用 'doc'）。
export type DocType = string

// 树节点：`GET /projects/{id}/documents` 已经嵌套返回 children。
export interface DocumentTreeNode {
  id: number
  projectId: number
  parentId: number | null
  title: string
  docType: DocType
  sortOrder: number
  archived: boolean
  createdBy: number
  createdAt: string
  updatedAt: string
  children: DocumentTreeNode[]
}

// 文档摘要（创建 / 详情里的 document 字段）。
export interface DocumentSummary {
  id: number
  projectId: number
  parentId: number | null
  title: string
  docType: DocType
  sortOrder: number
  archived: boolean
  createdBy: number
  createdAt: string
  updatedAt: string
}

// 详情里的结构块（万物皆块）。前端目前只用整篇 markdown（content），块信息保留以备后续。
export interface DocumentNode {
  id: number
  nodeType: string
  structOrder: number
  content: string
  editedAt: string
}

export interface DocumentDetail {
  document: DocumentSummary
  content: string
  nodes: DocumentNode[]
}

export interface DocumentPatch {
  title?: string
  parentId?: number | null
  sortOrder?: number
  archived?: boolean
}

// 列出项目下的文档树（已嵌套）。
export const listDocuments = (projectId: number) =>
  NewApiInstance.request<{ documents: DocumentTreeNode[] }>({
    url: `/projects/${projectId}/documents`,
    method: 'GET',
  })

// 新建文档；parentId 省略 = 根级。
export const createDocument = (projectId: number, data: { title: string; parentId?: number | null }) =>
  NewApiInstance.request<{ document: DocumentSummary }>({
    url: `/projects/${projectId}/documents`,
    method: 'POST',
    data,
  })

// 获取单篇文档详情（含整篇 markdown content）。
export const getDocument = (documentId: number) =>
  NewApiInstance.request<DocumentDetail>({
    url: `/documents/${documentId}`,
    method: 'GET',
  })

// 保存整篇 markdown 正文。
export const saveDocument = (documentId: number, content: string) =>
  NewApiInstance.request<DocumentDetail>({
    url: `/documents/${documentId}`,
    method: 'PUT',
    data: { content },
  })

// 局部更新元信息（重命名 / 移动 / 排序 / 归档）。移到根传 parentId:null。
export const patchDocument = (documentId: number, patch: DocumentPatch) =>
  NewApiInstance.request<DocumentDetail>({
    url: `/documents/${documentId}`,
    method: 'PATCH',
    data: patch,
  })

// 删除文档（子文档会上提到被删文档的父级）。
export const deleteDocument = (documentId: number) =>
  NewApiInstance.request<null>({
    url: `/documents/${documentId}`,
    method: 'DELETE',
  })
